docs/Driver.md

```diff
@@ -0,0 +1,311 @@
+linker, the compiler needs to accomplish two other tasks:
+
+- **Autolinking:** Swift object files encode information about what libraries
+  they depend on. On Apple platforms the linker can read this information
+  directly; on other platforms it's extracted using the
+  `swift-autolink-extract` helper tool. Of course the build system can also
+  provide manual link commands too.
+
+- **Debugging:** In addition to the usual DWARF format, interactive debugging
+  of a Swift program requires access to its module. This is supported via a
+  custom linker flag on Apple platforms and by the `swift-modulewrap` tool
+  elsewhere. Apple platforms also invoke `dsymutil` after linking to produce a
+  dSYM debugging archive, so that the program is still debuggable even once the
+  object files have been cleaned.
+
+
+## Output File Maps ##
+
+> There are three numbers in computer science: 0, 1, and N.
+
+The problem with running a separate frontend job for each input file is that it
+suddenly becomes much more complicated to determine where outputs go. If you're
+just going to link everything together at the end this isn't really a problem,
+but a lot of build systems just want the compiler to produce object files. If
+you want to use Swift's cross-file dependency tracking, or get Clang-style
+serialized diagnostics (.dia files) instead of just text output, those are also
+generated per-input-file. And some build systems (\*cough\*Xcode\*cough\*) want
+to individually track persisted output files, instead of dumping them into a
+temporary directory. Trying to specify the outputs for every input file on the
+command line would be awkward and ridiculous, so instead we use an _output file
+map._
+
+An output file map contains a JSON dictionary that looks like this:
+
+```json
+{
+  "/path/to/src/foo.swift": {
+    "object": "/path/to/build/foo.o",
+    "swift-dependencies": "/path/to/build/foo.swiftdeps",
+    "diagnostics": "/path/to/build/foo.dia",
+  },
+  "/path/to/src/bar.swift": {
+    "object": "/path/to/build/bar.o"
+    "swift-dependencies": "/path/to/build/bar.swiftdeps",
+    "diagnostics": "/path/to/build/bar.dia",
+  },
+  "": {
+    "swift-dependencies": "/path/to/build/main-build-record.swiftdeps",
+  }
+}
+```
+
+The build system is responsible for generating this file. The input file names
+don't have to be absolute paths, but they do have to match the form used in the
+invocation of the compiler. No canonicalization is performed. The
+"swift-dependencies" entries are required to perform incremental builds (see
+below). The "diagnostics" entries are only for your own use; if you don't need
+them they can be omitted.
+
+The output file map accepts other entries, but they should not be considered
+stable. Please stick to what's shown here.
+
+(It *would* be nice to have a way to get reasonable default behavior by
+pointing `-o` at a directory, or by adding an `-intermediate-directory`
+argument or something. Patches welcome.)
```

Comment:
Do these niceties come with Swift JIRA tasks? Linking to those would make it (a) easy for contributors to start that work, and (b) easy to understand when this blurb can be deleted from this doc.